### PR TITLE
improvements and refactors

### PR DESCRIPTION
**Pull request by benjaglobant**
In this PR I implemented improvements and fixes in the application code, in order to have good code quality and generating a good user experience when interacting with the app.
 - Added messages of error more descriptive (SpellsActivity, HouseDetailFragment, CharactersFragment)
 - Used the analyzer to inspect the project's code and made the refactors that it suggest or add some ignores.
 - Tests refactor
 - MainActivity and MainViewModel refactor, now the app is fulfilling the MVVM in a good way in this feature.
 - Created a Constants object to have not the same constant declared in differents files.

Gif showing the error messages added:
improvementsPR

First analizer inspection:
image
Code inspection before resolve the warnings/errors:
image
Tests image:
image
Coverage image (ViewModels and UseCases):
coverage

### CODE REVIEW

File: app/src/main/res/layout/activity_splash.xml [hunk 1]
```diff
@@ -7,13 +7,15 @@
...
+    tools:ignore="Overdraw">
 
     <ImageView
         android:id="@+id/logo"
         android:layout_width="match_parent"
         android:layout_height="match_parent"
         android:layout_margin="@dimen/splash_activity_hogwarts_image_view_margin"
-        android:src="@drawable/hogwarts" />
+        android:src="@drawable/hogwarts"
+        tools:ignore="ContentDescription" />
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_splash.xml]: instead of ignore the warning, add the contentDescription field

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -11,7 +12,8 @@
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
-        android:theme="@style/Theme.AppCompat.NoActionBar">
+        android:theme="@style/Theme.AppCompat.NoActionBar"
+        tools:ignore="AllowBackup">
```
Comment (SartiMau) [inline app/src/main/AndroidManifest.xml]: is really necessary that you have the allowBackup in true?

Comment (SartiMau) [reply app/src/main/AndroidManifest.xml]: if not, please remove the ignore

File: app/src/main/java/com/globant/harrypotterapp/activity/MainActivity.kt [hunk 3]
```diff
@@ -18,33 +23,66 @@ class MainActivity : AppCompatActivity() {
...
+        startActivity(HouseActivity.getIntent(this, houseName))
+    }
+
+    private fun openSpellsActivity() {
+        startActivity(SpellsActivity.getIntent(this))
+    }
+
     companion object {
         fun getIntent(context: Context): Intent = Intent(context, MainActivity::class.java)
         private const val GRYFFINDOR = "Gryffindor"
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/activity/MainActivity.kt]: you have a constantUtils with this value, maybe you can put all the hases names there

File: app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt [hunk 4]
```diff
@@ -76,6 +79,7 @@ class CharactersFragment : Fragment(), OnCharacterClicked {
     }
 
     companion object {
+        private const val CHARACTERS = "Characters"
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt]: why did you need this? you are only using it in the line 68 in a getString. put this in the same resource

File: app/src/main/java/com/globant/harrypotterapp/fragment/HouseDetailFragment.kt [hunk 5]
```diff
@@ -51,20 +52,20 @@ class HouseDetailFragment : Fragment() {
...
+            houseDetailFragmentError.visibility = View.VISIBLE
+            houseDetailFragmentError.text = getString(R.string.error_message_text, message, HOUSE_DETAILS)
+        }
     }
 
     companion object {
-        private const val ZERO = 0
-        private const val GRYFFINDOR = "Gryffindor"
+        private const val HOUSE_DETAILS = "House details"
         private const val RAVENCLAW = "Ravenclaw"
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/fragment/HouseDetailFragment.kt]: same, put the houses names in the constant utils

File: app/src/main/java/com/globant/harrypotterapp/viewmodel/MainViewModel.kt [hunk 6]
```diff
@@ -16,28 +15,42 @@ import kotlinx.coroutines.withContext
...
+
+    override fun goToSpells() {
+        housesMutableLiveData.value = Event(MainData(status = MainStatus.GO_SPELLS))
+    }
 }
 
-data class MainData<RequestData>(var status: MainStatus, var data: RequestData? = null, var error: Exception? = null)
+data class MainData(var status: MainStatus, var houseName: String = EMPTY_STRING, var error: Exception? = null) {
+    companion object {
+        private const val EMPTY_STRING = ""
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/viewmodel/MainViewModel.kt]: put this in a constant utils

File: app/src/main/res/layout/activity_character_detail.xml [hunk 7]
```diff
@@ -30,7 +31,8 @@
                 android:layout_height="wrap_content"
                 android:scaleType="centerCrop"
                 android:src="@drawable/character_detail_image"
-                app:layout_collapseMode="parallax" />
+                app:layout_collapseMode="parallax"
+                tools:ignore="ContentDescription" />
```
Comment (SartiMau) [inline app/src/main/res/layout/activity_character_detail.xml]: add the content description instead of the ignore

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt [hunk 8]
```diff
@@ -83,12 +88,18 @@ class CharacterDetailActivity : AppCompatActivity() {
...
+            characterDetailActivityToolbar.title = getString(R.string.error_title)
+            characterDetailActivityError.visibility = View.VISIBLE
+            characterDetailActivityError.text =
+                getString(R.string.error_message_text, message, CHARACTER_DETAILS)
+        }
     }
 
     companion object {
         private const val CHARACTER_ID = "characterId"
+        private const val CHARACTER_DETAILS = "Character details"
```
Comment (SartiMau) [inline app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt:102]: you are using this in the getString on line 96, put this in the same string res

File: app/src/main/res/layout/fragment_house_detail.xml [hunk 9]
```diff
@@ -19,14 +20,27 @@
...
         <ImageView
             android:id="@+id/house_detail_fragment_image"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             app:layout_constraintEnd_toEndOf="parent"
             app:layout_constraintStart_toStartOf="parent"
             app:layout_constraintTop_toTopOf="parent"
-            android:layout_marginTop="@dimen/house_detail_fragment_view_margin_top"/>
+            android:layout_marginTop="@dimen/house_detail_fragment_view_margin_top"
+            tools:ignore="ContentDescription" />
```
Comment (SartiMau) [inline app/src/main/res/layout/fragment_house_detail.xml]: add the content description instead of the ignore

File: domain/src/main/java/com/globant/domain/usecase/implementation/GetHousesUseCaseImpl.kt [hunk 10]
```diff
@@ -18,7 +18,7 @@ class GetHousesUseCaseImpl(
                 result
             }
             is Result.Failure -> {
-                database.getHouses()
+                result
```
Comment (SartiMau) [inline domain/src/main/java/com/globant/domain/usecase/implementation/GetHousesUseCaseImpl.kt:21]: are you sure of this? if the service fails, this may call the database I think

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/activity_splash.xml (hunk 1 above)
Comment (benjaglobant) [reply app/src/main/res/layout/activity_splash.xml]: Done!

File: app/src/main/java/com/globant/harrypotterapp/activity/MainActivity.kt (hunk 3 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/activity/MainActivity.kt]: You are right, done!

File: app/src/main/java/com/globant/harrypotterapp/fragment/HouseDetailFragment.kt (hunk 5 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/fragment/HouseDetailFragment.kt]: Done!

File: app/src/main/java/com/globant/harrypotterapp/viewmodel/MainViewModel.kt (hunk 6 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/viewmodel/MainViewModel.kt]: Done!

File: app/src/main/res/layout/activity_character_detail.xml (hunk 7 above)
Comment (benjaglobant) [reply app/src/main/res/layout/activity_character_detail.xml]: Done!

File: app/src/main/res/layout/fragment_house_detail.xml (hunk 9 above)
Comment (benjaglobant) [reply app/src/main/res/layout/fragment_house_detail.xml]: Done!

File: domain/src/main/java/com/globant/domain/usecase/implementation/GetHousesUseCaseImpl.kt (hunk 10 above)
Comment (benjaglobant) [reply domain/src/main/java/com/globant/domain/usecase/implementation/GetHousesUseCaseImpl.kt:21]: Yes I am sure, because I get the houses from database in other use case, in this one I just need to get the houses from the service and update the houses in database.

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (benjaglobant) [reply app/src/main/AndroidManifest.xml]: Refactored to "false", thanks Mau!

File: app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt (hunk 4 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/fragment/CharactersFragment.kt]: Refactored to string resource.

File: app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt (hunk 8 above)
Comment (benjaglobant) [reply app/src/main/java/com/globant/harrypotterapp/activity/CharacterDetailActivity.kt:102]: Refactored to string resource

**Review by SartiMau — APPROVED** [review]
